Adjustable Field Size

ShadowProtocol opened the pull request:
Implements #1644

# Changes proposed in this pull request:
- The scarecrow field GUI has four new directional buttons
    - Tooltips help the player orientate themselves
- Buttons in scarecrow GUI control how large the worked plot size of the field is (buttons toggle 0 through 5)
- Farmer AI (EntityAIWorkFarmer) will skip over blocks outside the defined size without requiring a new worker tick each time

The new GUI (image 1) and neatly sized plot (image 2)
![2020-07-14_21 52 22](https://user-images.githubusercontent.com/30493222/87424210-ec0fd300-c61e-11ea-924d-50f3ac90896c.png)
![2020-07-14_13 38 43](https://user-images.githubusercontent.com/30493222/87424309-13ff3680-c61f-11ea-9345-02f7a6014e4f.png)

Review please

ShadowProtocol committed:
Add field size buttons

Correct spelling mistake in class name AbstractScarescrowTileEntity
Add Custom widgets for size buttons

ShadowProtocol committed:
Fix farmer AI to include changes

Adds a new method for finding the next available block that considers the field plot size
Fixes the server syncing with the buttons in WindowField

ShadowProtocol committed:
Add localized tooltips for scarecrow size buttons

Show the direction of the button in a tooltip
And the relative direction from the player

ShadowProtocol committed:
Localize and clean up

Adds new I18n keys according to the correct convention
Removes pointless name variable and associated getters in ScarecrowTileEntity
Add some JavaDocs

ShadowProtocol committed:
Clean up imports

Raycoms commented:
Why doesn't it say the worker name anymore that is assigned to it?

Nightenom commented on src/main/resources/assets/minecolonies/lang/en_us.json:
```diff
@@ -432,6 +432,15 @@
   "com.minecolonies.coremod.gui.buildtool.hut.level": "Level %s",
   "com.minecolonies.coremod.gui.buildtool.decorations": "Decorations",
   "com.minecolonies.coremod.gui.buildtool.scans": "My schematics",
+  "gui.field.north": "Northward field size",
```
use modid in lang keys

ShadowProtocol replied:
Missed this before somehow. Anyways adding "com" and "coremod" is adding megabytes onto the lang files all up, and is not really current convention. So do you want "gui.minecolonies.field..." instead?
Also quite happy to do a PR to tidy up all these lang files (removing deprecation, sort in order, etc). Not that hard with project-wide find-and-replace.

Nightenom replied:
that's up to you, ideally, it should start/contain modid to prevent collisions with everything else

Nightenom replied:
btw do not sort them, and we should be adding them to the end of the file, removing is fine, beware of project replace as it usually replaces more than you need :D

ShadowProtocol replied:
Alright then, gui.minecolonies it is. And I never use project replace without regex, to be ultra careful :)

Raycoms replied:
Make it the default com.minecolonies.coremod.gui, those few extra bytes in exchange for consistency. Maybe we rename them in the future to a shorter one. But for now that's the full gradle modid.

ShadowProtocol replied:
Not that gradle really has anything to do with the games I18n at all. I'll put them with the full thing then for now, but I would like to change them all in another tidy up PR if there's no big objections. It has no advantages being like this, and to me looks kinda messy. (and if someone's offering to clean up for free, wouldn't you take it? :D)

Raycoms replied:
We can clean those up when we drop 1.15 support. If you do this now, this will lead to massive conflicts with 1.16 potentially =D

ShadowProtocol replied:
Sure, that's fair.

Raycoms committed:
Merge branch 'version/1.15' into feature/field

ShadowProtocol commented:
> Why doesn't it say the worker name anymore that is assigned to it?

It does. The name will appear above the GUI once ScarecrowTileEntity#setOwner is called (I have not changed when setOwner gets called afaik). If there is no farmer, there is no name. Did you want a "No farmer assigned" placeholder?

Raycoms commented:
Nvm, I didn't see it on top of it. The position changed and I got confused xDD

ShadowProtocol committed:
Add MOD_ID to translation keys

someaddons commented on src/main/java/com/minecolonies/coremod/client/gui/containers/WindowField.java:
```diff
@@ -40,7 +55,9 @@
-    private final AbstractScarescrowTileEntity tileEntity;
+    private final AbstractScarecrowTileEntity tileEntity;
+
+    private final int[] radii = new int[4];
```
javadoc

someaddons commented on src/main/java/com/minecolonies/coremod/client/gui/containers/WindowField.java:
```diff
@@ -53,18 +70,66 @@ public WindowField(final ContainerField container, final PlayerInventory playerI
+    {
+        super.init();
+
+        final int offset = 24;
```
make a constant for this

someaddons commented on src/main/java/com/minecolonies/coremod/client/gui/containers/WindowField.java:
```diff
@@ -91,4 +156,118 @@ public void render(int x, int y, float z)
         this.renderHoveredToolTip(x, y);
     }
+
+    protected class DirectionalButton extends Button
```
javadoc

someaddons reviewed (dismissed):
Looks good, just minor doc/lang key prefixes missing

ShadowProtocol committed:
Add docs and append translation keys

Move translation keys from amidst the file to the end.
Make a static constant for button width

ShadowProtocol committed:
Revert translation keys

Introduce the formal mod id back into the scarecrow translation keys

ShadowProtocol committed:
Fix tooltip relative directions

ShadowProtocol committed:
Merge branch 'feature/field' of https://github.com/ShadowProtocol/minecolonies into feature/field

 Conflicts:
	src/main/resources/assets/minecolonies/lang/en_us.json

marchermans reviewed (approved):
Somebody file fixed the damn Scarescrow typo.
:D
